### DISCUSSION TITLE
SQLITE_READONLY: attempt to write a readonly database

### DISCUSSION THREAD

**Discussion by xaqbr** [pockethost/pockethost#262] [Q&A]
I recently created an instance and uploaded my pb_migrations and pb_data folder to it through FTP.

But I am getting an error now every time I connect to it: https://rpchat.pockethost.io/

Removing everything from the folders doesn't seem to work either, and I also tried changing versions through the Danger Zone but that didn't help.

**Comment by benallfree** [comment]
@wayjack What happens if you create a new instance?

**Comment by jramke** [comment]
Had the same problem, new instance is working.
But is there a way you can delete instances?

**Comment by benallfree** [comment]
We don't have a way to delete instances, but @jramke @wayjack I'd like to get more info from each of you. I wonder if there is a problem with uploading database backups. That `read-only` issue is likely something wrong with the system, I'd like to track it down.

One potential issue is that uploading should really only be done when the instance is in maintenance mode and not running. Otherwise I'm not sure what happens when you overwrite an active/open database file.

**Reply by jramke** [reply]
Oh yeah i didnt consider the maintance mode 😅. I think that causes the error.

Im going to try it again in maintance mode when i have time.

**Comment by BasToTheMax** [comment]
Could be database corruption?

**Comment by jramke** [comment]
@benallfree everything working as aspected when uploading pb_data and pb_ migrations folders via FTP when the maintance mode is activated.

**Comment by infiliset** [comment]
Hi all! I've got the same problem here: https://tablepay.pockethost.io/
Is there a workaround to solve the issue without creating new instances? Thanks!

**Reply by benallfree** [reply]
Interesting, I hope we can get to the bottom of this. Is your instance bricked right now?

**Reply by infiliset** [reply]
Exactly... even tried to re-upload pb_data's but nothing changed. Any suggestions?
I think all started when I accidentally deleted pb_data's files, but I thought was enough re-uploading good ones

**Reply by benallfree** [reply]
I wonder if uploading is changing things to read-only. Let me check on that.

**Reply by benallfree** [reply]
It doesn't appear to be a file permission issue. Did you see this? https://github.com/benallfree/pockethost/discussions/262#discussioncomment-6555995

**Reply by infiliset** [reply]
What you are suggesting is to put the instance in maintenance mode, re-uploading pb_data's files and trying again? Correct? I think I did just like that but I'm not 100% sure, if necessary I can retry... Thanks again

**Reply by benallfree** [reply]
Yes, please see if that works for you.

**Reply by infiliset** [reply]
I tried but I've got the same error: "SQLITE_READONLY: attempt to write a readonly database"
I put the instance in maintenance mode, deleted the data.db and logs.db and uploaded fresh ones taken from local computer but nothing changed.
So strange...

**Reply by infiliset** [reply]
@benallfree , my host now is working! yeah! Did you change something? No more SQLITE_READONLY error!

**Reply by benallfree** [reply]
Didn't touch a thing! How strange!

**Reply by benallfree** [reply]
@ganigeorgiev do you have any thoughts on this mystery? Have you seen `SQLITE_READONLY: attempt to write a readonly database` in your testing?

**Reply by ganigeorgiev** [reply]
If it's not a permissions issue, then I'm not sure what could be the cause of the above error. Reading briefly the above discussion it seems that it sometimes happens when deleting `pb_data` while the application is still running.

Note that on unix systems a file will not be truly deleted if there is at least 1 open descriptor to it because the related inode will be still in use. This is also the reason why for example you can run `rm` on the executable while still running and upload another one in its place. But the new executable will not have effect until the process is restarted.

I'm not familiar how the maintenance mode works, but my guess is that when uploading a new `pb_data` the application is not restarted and it still references the old db file (_which probably after the deletion is moved internally to a temp location where the current user don't have write access?_).

**Reply by benallfree** [reply]
@ganigeorgiev  Maintenance mode closes the pocketbase process and prohibits it from starting. I had the same thought about file descriptors being left open. Thanks for brainstorming!

I may update the FTP proxy to prohibit writes to `pb_data/*` unless the instance is in maintenance mode and the pocketbase process has exited.

**Reply by infiliset** [reply]
Thank you so much guys! I love PocketBase and PocketHost, and can't wait to see the v0.17 running on PH with JS hooks! :-) Keep It up

**Comment by zhaivoronok** [comment]
Same problem here: https://teploteh-admin.pockethost.io.

I tried to upload my local backup through the backup upload interface in admin settings and it works, but after a few minutes the error appeared.
I have restored the initial backup that I've made right after the instance creation but that didn't help.

**Reply by benallfree** [reply]
Is this still happening? The original thread was about people uploading through SFTP, but you're referring to restoring from the admin backup tool?
